contrib/libpcap/utils/utils_libpcap.cc

```diff
@@ -0,0 +1,170 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "contrib/libpcap/utils/utils_libpcap.h"
+
+#include <fstream>
+#include <iostream>
+#include <string>
+
+#include "contrib/libpcap/sandboxed.h"
+
+constexpr size_t kMaxPacket = 1024 * 1024 * 128;  // 128MB
+
+absl::Status LibPcap::OpenRemote() {
+  SAPI_RETURN_IF_ERROR(CheckIsInit());
+
+  SAPI_RETURN_IF_ERROR(sandbox_->TransferToSandboxee(&fd_));
+
+  sapi::v::CStr mode("r");
+  SAPI_ASSIGN_OR_RETURN(void* pfile,
+                        api_.sapi_fdopen(fd_.GetRemoteFd(), mode.PtrBefore()));
+  sapi::v::RemotePtr pcap_pfile(pfile);
+
+  sapi::v::GenericPtr sapi_pcap_errmsg;
+  SAPI_ASSIGN_OR_RETURN(
+      pcap_t * pcap_context,
+      api_.pcap_fopen_offline(&pcap_pfile, sapi_pcap_errmsg.PtrAfter()));
+
+  if (pcap_context == nullptr) {
+    api_.sapi_fclose(&pcap_pfile).IgnoreError();
+    sapi::v::RemotePtr remote_ptr_error(sapi_pcap_errmsg.GetRemote());
+    SAPI_ASSIGN_OR_RETURN(std::string errmsg,
+                          sandbox_->GetCString(remote_ptr_error));
+    return absl::UnavailableError(errmsg);
+  }
+
+  sapi_pcap_context_.SetRemote(pcap_context);
+  return absl::OkStatus();
+}
+
+LibPcap::~LibPcap() {
+  if (sapi_pcap_context_.GetRemote() != nullptr) {
+    api_.pcap_close(sapi_pcap_context_.PtrNone()).IgnoreError();
+  }
+}
+
+bool LibPcap::IsInit() { return CheckIsInit().ok(); }
+
+absl::Status LibPcap::CheckIsInit() {
+  if (fd_.GetValue() < 0) {
+    return absl::UnavailableError("PCAP file not opened");
+  }
+  if (!init_status_.ok()) {
+    return init_status_;
+  }
+
+  return absl::OkStatus();
+}
+
+absl::Status LibPcap::GetInitStatus() { return init_status_; }
+
+absl::StatusOr<LibPcapPacket> LibPcap::Next() {
+  SAPI_RETURN_IF_ERROR(CheckIsInit());
+
+  LibPcapPacket packet(&sapi_pcap_context_, &api_);
```

Comment:
`FetchPackage` (s/Package/Packet?) is already de facto factory method.
Just make it a proper static one that returns `absl::StatusOr<LibPcapPacket>`
Also `Finished()` should not be needed then as it could just return an `absl::OutOfRangeError`

Reply:
Fixed naming.
I have removed Finished function.

That said I'm not sure if I follow the idea of factory method.
This function returns an object based on the file, filter ect. defined in the parent class. So is this a proper factory?

Reply:
The problem I have with current state is `LibPcapPacket` mixes 2 things:
 - state of the fetching operation: is it past-the-end? (finished) did fetch fail? (non-ok init_status)
 - the pakcet data itself
Code that takes such object has to either verify it's state (`!Finished() && GetInitStatus.ok()`) or just implicitly assume it.